### ISSUE TITLE
Decode the TS Bank

### ISSUE THREAD

**Issue by rafopar** [JeffersonLab/hps-java#532]
We need to decode the TS bank, and store trigger bits
that are set in the TS bank.

**Comment by mholtrop** [comment]
This will need a completely new class, since the current trigger information is completely different from the 2015/16 data, and trying to make a driver that does both will be horrible.

Digging into this, and with help from Ben Raydo, I gather the following:

Hardware is described in: https://coda.jlab.org/drupal/system/files/pdfs/HardwareManual/TS/TS.pdf 
This is some serious engineer lingo filled cryptic stuff that makes you really appreciate we have excellent engineers at JLab. The important stuff for us is in the appendix: page 89-90.

We want to decode the 57610 bank that is inside the "46" bank, since only that one contains all the trigger bits. There are 7 uint32 words in the bank. 
A C++ decoder is here: https://github.com/JeffersonLab/EvioTool/blob/master/HPSEvioReader/TIBank.h 
(in process of being updated.)

     Meaning of the words for 2019 data: 

     0: (Event word 1 = Header )  <== Not as useful as is sounds. Use Event word 5 instead!
        Bit(31:24): Trigger Type;
        0: filler events,
        1-32: GTP Physics trigger;
        33-64: Front panel physics trigger;
        250: multi-hit on GTP or Front panel physics trigger 251: Multi-hit on GTP and Front panel physics   trigger 253: VME trigger;
        254: VME random trigger;
        For TImaster bit(31:26) represents the TS#6-1 inputs;
        Bit(23:16): 0000,0001, or 0x01;
        Bit(15:0): Event wordcount; Event header is excluded from the count

     1: (Event word 2) = Trigger number (low 32 bits) = Event number.
     2: (Event word 3) = Trigger timing 4ns steps (low 32 bits)
     3: (Event word 4) = bits 31:16 - high 16 bits of trigger number, bits 15:0 - high 16 bits of trigger timing.
     4: (Event word 5) = VTP Trigger bits AFTER prescaling is applied.
     5: (Event word 6) = EXT Trigger bits AFTER prescaling is applied.
     6: (Event word 7) = bits 31:16 EXT, bits 15:0  VTP trigger bits before TS prescaling is applied.
          Note that HPS prescales mostly in the VTP, so the low bits in 4,5 will repeat in 6, EXCEPT for the Faraday cup trigger.

     The trigger bits are defined in the trigger configuration file and can be found in the VTP_HPS_PRESCALE entries where the rescale numbers are set. This is also in the tag="17" event at the start of the run.

     VTP_HPS_PRESCALE               0        0   # Single 0 Top    ( 150-8191) MeV (-31,31)   Low energy cluster
     VTP_HPS_PRESCALE               1        0   # Single 1 Top    ( 300-3000) MeV (  5,31)   e+
     VTP_HPS_PRESCALE               2        0   # Single 2 Top    ( 300-3000) MeV (  5,31)   e+ : Position dependent energy cut
     VTP_HPS_PRESCALE               3        0   # Single 3 Top    ( 300-3000) MeV (  5,31)   e+ : HODO L1*L2  Match with cluster
     VTP_HPS_PRESCALE               4        0   # Single 0 Bot    ( 150-8191) MeV (-31,31)   Low energy cluster

     VTP_HPS_PRESCALE               5        0   # Single 1 Bot    ( 300-3000) MeV (  5,31)   e+
     VTP_HPS_PRESCALE               6        0   # Single 2 Bot    ( 300-3000) MeV (  5,31)   e+ : Position dependent energy cut
     VTP_HPS_PRESCALE               7        0   # Single 3 Bot    ( 300-3000) MeV (  5,31)   e+ : HODO L1*L2  Match with cluster
     VTP_HPS_PRESCALE               8        0   # Pair 0          A'
     VTP_HPS_PRESCALE               9        0   # Pair 1          Moller
     VTP_HPS_PRESCALE               10       0   # Pair 2          pi0
     VTP_HPS_PRESCALE               11       0   # Pair 3          -
     VTP_HPS_PRESCALE               12       0   # LED
     VTP_HPS_PRESCALE               13       0   # Cosmic
     VTP_HPS_PRESCALE               14       0   # Hodoscope
     VTP_HPS_PRESCALE               15       0   # Pulser
     VTP_HPS_PRESCALE               16       0   # Multiplicity-0 2 Cluster Trigger
     VTP_HPS_PRESCALE               17       0   # Multiplicity-1 3 Cluster trigger
     VTP_HPS_PRESCALE               18       0   # FEE Top       ( 2600-5200)
     VTP_HPS_PRESCALE               19       0   # FEE Bot       ( 2600-5200)

**Comment by rafopar** [comment]
Hello Maurik, just for the sake of not doubling the effort,
Tongtong has started to work on this,

in particular he already implemented the item number 4
` 4: (Event word 5) = VTP Trigger bits AFTER prescaling is applied.`

I just added him in the assignee list, that efforts will be more coordinated

**Comment by mholtrop** [comment]
Hello Rafo, Tongtong, that is good. And yes, always add your name if you are working on an issue!!!!

For this situation, I think you should make a new class, say TIData2019, instead of changing the current TIData.java class. You are too likely to break 2015/2016 analysis if you keep the TIData class and modify it.

I am not sure why we need: evio/src/main/java/org/hps/evio/TSGenericObject.java. That looks like clutter to me, but I could be wrong. 

Also, I disagree with this type of coding: (https://github.com/JeffersonLab/hps-java/compare/iss532#diff-9016c39827f3ff8cae5c9ba6d8b0a777R58)
```java
                            // the trigger pattern is in the 5th word
                            for (int i = 0; i < 32; i++) {
                                if (vals[4] % 2 == 1) {
                                    list.add(i);
                                }
                                vals[4] = vals[4] / 2;
                            }
                            Integer[] array = list.toArray(new Integer[0]);
                            int[] triggers = new int[array.length];
                            for (int i = 0; i < array.length; i++) {
                                triggers[i] = array[i].intValue();
                            }
                            trigger_flags = ArrayUtils.addAll(trigger_flags, triggers);
```

The only thing you are doing here is unpacking a 32-bit integer into 32 integers, thus increasing the space requirements by a factor of 32 and not adding anything new, nor helping anyone decode the content. DON'T!  Yes, LCIO is too stupid to contain an array of booleans, and Java is too stupid to contain unsigned integers, but we still don't want to do this.

It does appear that Java does not contain the bitwise structure of C, however it does have "BitSet" (https://docs.oracle.com/javase/7/docs/api/java/util/BitSet.html). 

So store the whole word into the generic object, and then provide a class that unpacks that word and makes the individual bits accessible. 

Note that the Trigger Time (which is in 4ns increments) is a useful number to store in the generic object as well.

**Comment by tongtongcao** [comment]
Hello Maurik,

I’d like to explain what the code does.
The code picks up bits, which values are 1, and stores locations of bits into an integer array.
The elements in the array for each event match with # in the table of https://logbooks.jlab.org/system/temporary/clonpc14_20190628_154354.gif <https://logbooks.jlab.org/system/temporary/clonpc14_20190628_154354.gif>

For example, the trigger patten is 0000,0000,0000,0000,0000,0000,0100,0100, then the 2nd & 6th bits (starting from 0) are picked up, and 2 & 6 are stored into an integer array.
Check the list in https://logbooks.jlab.org/system/temporary/clonpc14_20190628_154354.gif <https://logbooks.jlab.org/system/temporary/clonpc14_20190628_154354.gif>, 2 & 6 correspond Single-2 top and Single-2 bot.

Best regards,
Tongtong

**Comment by rafopar** [comment]
Hello Maurik, 

Yes, I wanted to assign Tongtong's name to it, but at that time he was not a part of JeffersonLab Group, and I wasn't able to do it at that time.

About `Also, I disagree with this type of coding`, that was my idea to do it as a generic object.
I clearly understand it might sound like we are increasing the volume x32, however, more than 90-95% of the time TS trigger word has only one bit set (unlike to VTP banks, where all trigger decisions are saved), 
and  few percent of events have 2. 3 and more bits are extremely rare, which means 95% of the time the genericObject will have only one number.

My thought, that generic object is "better", wrt providing a class, is that the class is usable in java only, but if someone works in C++ API to convert data to root, then this probably will not work (Maybe I am wrong, and should have this be discussed with a software group, before making suggestions to Tongtong)

**Comment by mholtrop** [comment]
OK, you are correct, I misread. The code now translates 0x00008000 to an array of ints: {15}, and 0x0008002 to an array {15,2} etc. 
I don’t see how this is a correct interpretation. Perhaps this is what you like in your C++ analysis, but I don’t think this is at all useful.

In C++ what you _should_ do is:
```c++
  struct TriggerBits {
    bool Single_0_Top: 1; //  0   ( 150-8191) MeV (-31,31)   Low energy cluster
    bool Single_1_Top: 1; //  1   ( 300-3000) MeV (  5,31)   e+
    bool Single_2_Top: 1; //  2   ( 300-3000) MeV (  5,31)   e+ : Position dependent energy cut
    bool Single_3_Top: 1; //  3   ( 300-3000) MeV (  5,31)   e+ : HODO L1*L2  Match with cluster
    bool Single_0_Bot: 1; //  4   ( 150-8191) MeV (-31,31)   Low energy cluster
    bool Single_1_Bot: 1; //  5   ( 300-3000) MeV (  5,31)   e+
    bool Single_2_Bot: 1; //  6   ( 300-3000) MeV (  5,31)   e+ : Position dependent energy cut
    bool Single_3_Bot: 1; //  7   ( 300-3000) MeV (  5,31)   e+ : HODO L1*L2  Match with cluster
    bool Pair_0      : 1; //  8    A'
    bool Pair_1      : 1; //  9    Moller
    bool Pair_2      : 1; // 10    pi0
    bool Pair_3      : 1; // 11    -
    bool LED         : 1; // 12    LED
    bool Cosmic      : 1; // 13    Cosmic
    bool Hodoscope   : 1; // 14    Hodoscope
    bool Pulser      : 1; // 15    Pulser
    bool Mult_0      : 1; // 16    Multiplicity-0 2 Cluster Trigger
    bool Mult_1      : 1; // 17    Multiplicity-1 3 Cluster trigger
    bool FEE_Top     : 1; // 18    FEE Top       ( 2600-5200)
    bool FEE_Bot     : 1; // 19    FEE Bot       ( 2600-5200)
    unsigned int NA  :12; // 20-31 Not used
  };
```
And then something like:
```c++
   unsigned int trigger_int = get_my_data_from_where_ever_function();
   TriggerBits *trigger_bits = reinterpret_cast<TriggerBits *>(&trigger_int);
   it( trigger_bits->Hodoscope ){  // Do stuff for the hodoscope }
```
If you are going to use C++, use it properly. Looping over some array to see if there is a value somewhere that is set to 14 is just inefficient.

In Java, you could use a BitSet, but now you still need some decoder class, cause Java ain't got no bitwise structures, and you need to specify somehow which bit is which.

**Comment by mholtrop** [comment]
Merged in #546
